library/core/class.session.php

```diff
@@ -132,42 +134,38 @@ public function checkPermission($Permission, $FullMatch = true, $JunctionTable =
 
-        $Permissions = $this->getPermissions();
-        if ($JunctionTable && !c('Garden.Permissions.Disabled.'.$JunctionTable)) {
-            // Junction permission ($Permissions[PermissionName] = array(JunctionIDs))
+        if ($JunctionID === '') {
+            $JunctionID = null;
+        }
+
+        $allPermissions = $this->permissions->getPermissions();
+
+        // Junction (e.g. category) permissions
+        if ($JunctionTable && !c("Garden.Permissions.Disabled.{$JunctionTable}")) {
             if (is_array($Permission)) {
-                $Pass = false;
-                foreach ($Permission as $PermissionName) {
-                    if ($this->checkPermission($PermissionName, false, $JunctionTable, $JunctionID)) {
-                        if (!$FullMatch) {
-                            return true;
-                        }
-                        $Pass = true;
-                    } else {
-                        if ($FullMatch) {
-                            return false;
-                        }
-                    }
+                if ($FullMatch) {
+                    return $this->permissions->hasAll($Permission, $JunctionID);
+                } else {
+                    return $this->permissions->hasAny($Permission, $JunctionID);
                 }
-                return $Pass;
             } else {
-                if ($JunctionID !== '') {
-                    $Result = array_key_exists($Permission, $Permissions)
-                        && is_array($Permissions[$Permission])
-                        && in_array($JunctionID, $Permissions[$Permission]);
+                if ($JunctionID !== null) {
+                    return $this->permissions->has($Permission, $JunctionID);
                 } else {
-                    $Result = array_key_exists($Permission, $Permissions)
-                        && is_array($Permissions[$Permission])
-                        && count($Permissions[$Permission]);
+                    // Backwards-compatible support for a null JunctionID meaning "permission on any ID"
+                    return array_key_exists($Permission, $allPermissions) && count($allPermissions[$Permission]);
                 }
-                return $Result;
             }
         } else {
-            // Non-junction permission ($Permissions = array(PermissionNames))
             if (is_array($Permission)) {
-                return arrayInArray($Permission, $Permissions, $FullMatch);
+                if ($FullMatch) {
+                    return $this->permissions->hasAll($Permission);
+                } else {
+                    return $this->permissions->hasAny($Permission);
+                }
             } else {
-                return in_array($Permission, $Permissions) || array_key_exists($Permission, $Permissions);
+                // Backwards-compatible permission check
+                return $this->permissions->has($Permission) || array_key_exists($Permission, $allPermissions);
```

Comment:
I'd prefer this odd extra check to be removed. Can you not make Vanilla\Permissions handle this?